sidekick_core/lib/src/sidekick_context.dart

```diff
@@ -0,0 +1,206 @@
+
+  /// Caches expensive lookups
+  // static SidekickContextCache? _cache;
+
+  static SidekickContextCache _cache = _NoCache();
+
+  /// Returns the name of the CLI
+  static String get cliName {
+    return sidekickPackage.cliName;
+  }
+
+  // /// Returns the directory of the Flutter SDK sidekick uses for the [flutter] command
+  // ///
+  // /// This variable is usually set to a pinned version of the Flutter SDK per project, i.e.
+  // /// - https://github.com/passsy/flutter_wrapper
+  // /// - https://github.com/fluttertools/fvm
+  // ///
+  // /// The flutterSdkPath can be set in [initializeSidekick].
+  // static Directory? get flutterSdk => throw 'TODO';
+  //
+  // /// Returns the directory of the Dart SDK sidekick uses for the [dart] command
+  // ///
+  // /// Overrides the Dart SDK of [flutterSdk] when set
+  // ///
+  // /// The dartSdkPath can be set in [initializeSidekick].
+  // static Directory? get dartSdk => throw 'TODO';
+  //
+  // /// The main package/app which should be executed by the [flutter] command
+  // ///
+  // /// The mainProjectPath can be set in [initializeSidekick].
+  // ///
+  // /// It's optional, not every project has a mainProject, there are repositories
+  // /// with zero or multiple projects.
+  // static DartPackage? get mainProject => throw 'TODO';
+
+  /// The location of the sidekick package inside the [repository]
+  static Directory get sidekickPackageDir => sidekickPackage.root;
+
+  /// The sidekick package inside the [repository]
+  static SidekickPackage get sidekickPackage {
+    return _cache.getOrCreate('sidekickPackage', _findSidekickPackage);
+  }
+
+  static SidekickPackage _findSidekickPackage() {
+    final injectedPackageHome = env[_envPackageHome];
+    if (injectedPackageHome != null && injectedPackageHome.isNotBlank) {
+      // When called via entryPoint, immediately return the package
+      return SidekickPackage.fromDirectory(Directory(injectedPackageHome))!;
+    }
+
+    final script = File(DartScript.self.pathToScript);
+    final scriptPath = script.uri.path;
+    final pathSep = Platform.pathSeparator;
+
+    // When CLI is run with compiled entryPoint: <sidekick package>/build/cli.exe
+    if (scriptPath.endsWith('${pathSep}build${pathSep}cli.exe')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+
+    // When CLI is run with `dart bin/main.dart`: <sidekick package>/bin/main.dart
+    if (scriptPath.endsWith('${pathSep}bin${pathSep}main.dart')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+
+    // in `UpdateCommand` when the latest `update_sidekick_cli.dart` is written to <sidekick package>/build/update.dart to be executed
+    if (scriptPath.endsWith('${pathSep}bin${pathSep}update.dart')) {
+      return SidekickPackage.fromDirectory(script.parent.parent)!;
+    }
+
+    // Fallback strategy: searching all parent folders until we find a dart package root
+    final scriptDirectory = script.parent;
```

Comment:
We have to double-check that this fallback works when tests for sidekick commands work when they are executed in the users package.  `DartScript.self` then points to their test file, which leaves us without any information where our package is.

Reply:
`DartScript.self` is awesome and yes, it handles unit tests well :)